core/ops.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
+pub use crate::libdeno::OpId;
+use crate::PinnedBuf;
+use futures::Future;
+use std::collections::HashMap;
+
+pub type Buf = Box<[u8]>;
+
+pub type OpAsyncFuture<E> = Box<dyn Future<Item = Buf, Error = E> + Send>;
+
+pub(crate) type PendingOpFuture =
+  Box<dyn Future<Item = (OpId, Buf), Error = CoreError> + Send>;
+
+pub type OpResult<E> = Result<Op<E>, E>;
+
+pub enum Op<E> {
+  Sync(Buf),
+  Async(OpAsyncFuture<E>),
+}
+
+pub type CoreError = ();
+
+pub type CoreOp = Op<CoreError>;
+
+/// Main type describing op
+pub type CoreOpHandler = dyn Fn(&[u8], Option<PinnedBuf>) -> CoreOp;
+
+#[derive(Default)]
+pub struct OpRegistry {
+  pub ops: Vec<Box<CoreOpHandler>>,
+  pub op_map: HashMap<String, OpId>,
+}
+
+fn op_noop(_control: &[u8], _zero_copy_buf: Option<PinnedBuf>) -> CoreOp {
+  Op::Sync(Box::new([]))
+}
+
+fn op_get_op_map(op_registry: &OpRegistry) -> CoreOp {
+  let op_map = op_registry.get_op_map();
+  let op_map_json = serde_json::to_string(&op_map).unwrap();
+  let buf = op_map_json.as_bytes().to_owned().into_boxed_slice();
+  Op::Sync(buf)
+}
+
+impl OpRegistry {
+  pub fn new() -> Self {
+    let mut registry = Self::default();
+    // Add single noop symbolizing "get_op_map" function. The actual
+    // handling is done in `call_op` method.
+    let op_id = registry.register_op("get_op_map", Box::new(op_noop));
```

Comment:
Why not `register_op("get_op_map", op_get_op_map)` ?

Reply:
See [this comment](https://github.com/denoland/deno/pull/3002#discussion_r326864872). Can't pass reference to `OpRegistry` and emplace this reference in the same registry and make borrow checker happy.

Reply:
Hmm... I can try to fix this in a bit. It seems like we should figure this out now. It should be possible.

Reply:
could you move the Box::new into register_op?

Reply:
> could you move the Box::new into register_op?

I don't think so, compiler doesn't like passing unboxed functions